idaes/core/solvers/petsc.py

```diff
@@ -0,0 +1,699 @@
+        wsl (bool): if True use WSL to run PETSc, if False don't use WSL to run
+            PETSc, if None automatic. The WSL is only for Windows.
+        keepfiles (bool): pass to keepfiles arg for solvers
+        symbolic_solver_labels (bool): pass to symbolic_solver_labels argument
+            for solvers. If you want to read trajectory data from the
+            time-stepping solver, this should be True.
+        vars_stub (str or None): Copy the `*.col` and `*.typ` files to the
+            working directory using this stub if not None.  These are needed to
+            interpret the trajectory data.
+        trajectory_save_prefix (str or None): If a string is provided the
+            trajectory data will be saved as gzipped json
+
+    Returns:
+        List of solver results objects from each solve. If there are initial
+        condition constraints and they are not skipped, the first object will
+        be from the initial condition solve.  Then there should be one for each
+        time element for each TS solve.
+    """
+    solve_log = idaeslog.getSolveLogger("petsc-dae")
+    regular_vars, time_vars = flatten_dae_components(m, time, pyo.Var)
+    regular_cons, time_cons = flatten_dae_components(m, time, pyo.Constraint)
+    tdisc = find_discretization_equations(m, time)
+
+    solver_snes = pyo.SolverFactory("petsc_snes", options=snes_options, wsl=wsl)
+    solver_dae = pyo.SolverFactory(
+        "petsc_ts", options=ts_options, wsl=wsl, vars_stub=vars_stub
+    )
+
+    if initial_variables is None:
+        initial_variables = []
+    if initial_constraints is None:
+        initial_constraints = []
+
+    if detect_initial:
+        rvset = ComponentSet(regular_vars)
+        rcset = ComponentSet(regular_cons)
+        icset = ComponentSet(initial_constraints)
+        ivset = ComponentSet(initial_variables)
+        initial_variables = list(ivset | rvset)
+        initial_constraints = list(icset | rcset)
+
+    # First calculate the inital conditions and non-time-indexed constraints
+    res_list = []
+    t0 = time.first()
+    if not skip_initial:
+        with TemporarySubsystemManager(to_deactivate=tdisc):
+            constraints = [
+                con[t0] for con in time_cons if t0 in con
+            ] + initial_constraints
+            variables = [var[t0] for var in time_vars] + initial_variables
+            if len(constraints) > 0:
+                # if the initial condition is specified and there are no
+                # initial constraints, don't try to solve.
+                t_block = create_subsystem_block(
+                    constraints,
+                    variables,
+                )
+                # set up the scaling factor suffix
+                _sub_problem_scaling_suffix(m, t_block)
+                with idaeslog.solver_log(solve_log, idaeslog.INFO) as slc:
+                    res = solver_snes.solve(t_block, tee=slc.tee)
+        res_list.append(res)
```

Comment:
Should this be indented twice more? To be inside the `if len(constraints) > 0` branch?